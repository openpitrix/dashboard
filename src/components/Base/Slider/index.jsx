import { isFunction, clamp } from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import styles from './index.scss';

export default class Slider extends React.Component {
  static propTypes = {
    children: PropTypes.any,
    className: PropTypes.string,
    max: PropTypes.number,
    min: PropTypes.number,
    onChange: PropTypes.func,
    step: PropTypes.number,
    value: PropTypes.number
  };

  static defaultProps = {
    className: '',
    min: 0,
    max: 100,
    step: 1,
    value: 0
  };

  constructor(props) {
    super(props);

    this.state = {
      maxPos: 0
    };
  }

  componentDidMount() {
    if (!this.slider || !this.block) {
      return;
    }
    const sliderWidth = this.slider.offsetWidth;
    const blockWidth = this.block.offsetWidth;
    this.setState({
      maxPos: sliderWidth - blockWidth
    });
  }

  getValue = pos => {
    const { maxPos } = this.state;
    const { min, max, step } = this.props;

    if (!maxPos) return 0;

    const percent = clamp(pos, 0, maxPos) / maxPos;
    const value = Math.round(percent * (max - min) / step) * step + min;

    return value;
  };

  getPosition = value => {
    const { maxPos } = this.state;
    const { min, max } = this.props;

    return Math.round((value - min) / (max - min) * maxPos);
  };

  handleStart = () => {
    document.addEventListener('mousemove', this.handleMove);
    document.addEventListener('mouseup', this.handleEnd);
  };

  handleMove = e => {
    e.stopPropagation();

    const pos = e.clientX - this.slider.getBoundingClientRect().left;
    const value = this.getValue(pos);

    if (isFunction(this.props.onChange)) {
      this.props.onChange(value);
    }
  };

  handleEnd = () => {
    document.removeEventListener('mousemove', this.handleMove);
    document.removeEventListener('mouseup', this.handleEnd);
  };

  render() {
    const { className, value } = this.props;

    const position = this.getPosition(value);
    const barStyle = { width: position };
    const blockStyle = { left: position };

    return (
      <div
        ref={ref => {
          this.slider = ref;
        }}
        className={classnames(styles.slider, className)}
        onMouseDown={this.handleMove}
        onMouseUp={this.handleEnd}
      >
        <div className={styles.bar} style={barStyle} />
        <div
          ref={ref => {
            this.block = ref;
          }}
          className={styles.block}
          style={blockStyle}
          onMouseDown={this.handleStart}
        />
      </div>
    );
  }
}
