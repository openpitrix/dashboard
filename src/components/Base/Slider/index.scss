@import '~scss/vars';

.slider {
  position: relative;
  display: inline-block;
  width: 276px;
  height: 14px;
  border-radius: 2px;
  background-color: #ffffff;
  border: solid 1px $border-lighter;
  cursor: pointer;
  .bar {
    height: 100%;
    background-color: #d7c9e6;
  }

  .block {
    position: absolute;
    top: -50%;
    width: 14px;
    height: 24px;
    border-radius: 2px;
    background-color: #ffffff;
    box-shadow: 0 1px 4px 0 rgba(48, 14, 86, 0.18);
  }
}
