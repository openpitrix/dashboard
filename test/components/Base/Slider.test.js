import Slider from 'components/Base/Slider';

describe('Base/Slider', () => {
  it('basic render', () => {
    const onChange = jest.fn();
    const wrapper = render(<Slider onChange={onChange}/>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
