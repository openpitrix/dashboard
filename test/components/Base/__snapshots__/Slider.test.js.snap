// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Slider basic render 1`] = `
<div
  class="slider"
>
  <div
    class="bar"
    style="width:0"
  />
  <div
    class="block"
    style="left:0"
  />
</div>
`;
